File: modules/stringalgorithms/stringoperations/StringOperations.java
public class StringOperations {

    // Calculate string length manually
    public static int stringLength(String str) {
        int length = 0;
        for (char c : str.toCharArray()) {
            length++;
        }
        return length;
    }

    // Find first occurrence of a character in a string
    public static int findChar(String str, char ch) {
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == ch) {
                return i;
            }
        }
        return -1; // not found
    }

    // Reverse a string
    public static String reverseString(String str) {
        char[] chars = str.toCharArray();
        int n = chars.length;
        for (int i = 0; i < n / 2; i++) {
            char temp = chars[i];
            chars[i] = chars[n - 1 - i];
            chars[n - 1 - i] = temp;
        }
        return new String(chars);
    }

    // Convert string to uppercase
    public static String toUppercase(String str) {
        return str.toUpperCase();
    }

    // Check if two strings are equal
    public static boolean stringsEqual(String str1, String str2) {
        return str1.equals(str2);
    }

    // Simple pattern matching - find substring
    public static int findSubstring(String text, String pattern) {
        int textLen = text.length();
        int patternLen = pattern.length();

        for (int i = 0; i <= textLen - patternLen; i++) {
            boolean match = true;
            for (int j = 0; j < patternLen; j++) {
                if (text.charAt(i + j) != pattern.charAt(j)) {
                    match = false;
                    break;
                }
            }
            if (match) {
                return i;
            }
        }
        return -1;
    }

    // Count occurrences of a character
    public static int countChar(String str, char ch) {
        int count = 0;
        for (char c : str.toCharArray()) {
            if (c == ch) {
                count++;
            }
        }
        return count;
    }

    // Check if string is palindrome
    public static boolean isPalindrome(String str) {
        String cleaned = str.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        return cleaned.equals(reverseString(cleaned));
    }

    public static void main(String[] args) {
        System.out.println("=== String Operations in Java ===\n");

        // Test string length
        String str1 = "Hello, World!";
        System.out.println("String: '" + str1 + "'");
        System.out.println("Length: " + stringLength(str1) + "\n");

        // Test find character
        System.out.println("Find 'W' at index: " + findChar(str1, 'W'));
        System.out.println("Find 'Z' at index: " + findChar(str1, 'Z') + "\n");

        // Test string equality
        String str2 = "Hello, World!";
        String str3 = "Goodbye";
        System.out.println("'" + str1 + "' == '" + str2 + "': " + stringsEqual(str1, str2));
        System.out.println("'" + str1 + "' == '" + str3 + "': " + stringsEqual(str1, str3) + "\n");

        // Test substring finding
        System.out.println("Find 'World' in '" + str1 + "': index " + findSubstring(str1, "World") + "\n");

        // Test reverse
        String str4 = "REVERSE";
        System.out.println("Original: '" + str4 + "'");
        System.out.println("Reversed: '" + reverseString(str4) + "'\n");

        // Test uppercase
        String str5 = "lowercase";
        System.out.println("Original: '" + str5 + "'");
        System.out.println("Uppercase: '" + toUppercase(str5) + "'\n");

        // Test character count
        System.out.println("Count of 'l' in '" + str1 + "': " + countChar(str1, 'l') + "\n");

        // Test palindrome
        String pal1 = "A man, a plan, a canal: Panama";
        String pal2 = "Hello";
        System.out.println("Is '" + pal1 + "' palindrome: " + isPalindrome(pal1));
        System.out.println("Is '" + pal2 + "' palindrome: " + isPalindrome(pal2));
    }
}

File: modules/stringalgorithms/stringoperations/string_operations.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// Calculate string length manually
int string_length(const char *str) {
    int length = 0;
    while (str[length] != '\0') {
        length++;
    }
    return length;
}

// Find first occurrence of a character in a string
int find_char(const char *str, char ch) {
    int i = 0;
    while (str[i] != '\0') {
        if (str[i] == ch) {
            return i;
        }
        i++;
    }
    return -1; // not found
}

// Reverse a string in-place
void reverse_string(char *str) {
    int n = string_length(str);
    for (int i = 0; i < n / 2; i++) {
        char temp = str[i];
        str[i] = str[n - 1 - i];
        str[n - 1 - i] = temp;
    }
}

// Convert string to uppercase
void to_uppercase(char *str) {
    int i = 0;
    while (str[i] != '\0') {
        str[i] = toupper((unsigned char)str[i]);
        i++;
    }
}

// Check if two strings are equal
int strings_equal(const char *str1, const char *str2) {
    int i = 0;
    while (str1[i] != '\0' || str2[i] != '\0') {
        if (str1[i] != str2[i]) {
            return 0;
        }
        i++;
    }
    return 1;
}

// Simple pattern matching - find substring
int find_substring(const char *text, const char *pattern) {
    int text_len = string_length(text);
    int pattern_len = string_length(pattern);

    for (int i = 0; i <= text_len - pattern_len; i++) {
        int match = 1;
        for (int j = 0; j < pattern_len; j++) {
            if (text[i + j] != pattern[j]) {
                match = 0;
                break;
            }
        }
        if (match) {
            return i;
        }
    }
    return -1;
}

int main() {
    printf("=== String Operations in C ===\n\n");

    // Test string length
    char str1[] = "Hello, World!";
    printf("String: '%s'\n", str1);
    printf("Length: %d\n\n", string_length(str1));

    // Test find character
    printf("Find 'W' at index: %d\n", find_char(str1, 'W'));
    printf("Find 'Z' at index: %d\n\n", find_char(str1, 'Z'));

    // Test string equality
    char str2[] = "Hello, World!";
    char str3[] = "Goodbye";
    printf("'%s' == '%s': %s\n", str1, str2, strings_equal(str1, str2) ? "true" : "false");
    printf("'%s' == '%s': %s\n\n", str1, str3, strings_equal(str1, str3) ? "true" : "false");

    // Test substring finding
    printf("Find 'World' in '%s': index %d\n\n", str1, find_substring(str1, "World"));

    // Test reverse
    char str4[] = "REVERSE";
    printf("Original: '%s'\n", str4);
    reverse_string(str4);
    printf("Reversed: '%s'\n\n", str4);

    // Test uppercase
    char str5[] = "lowercase";
    printf("Original: '%s'\n", str5);
    to_uppercase(str5);
    printf("Uppercase: '%s'\n", str5);

    return 0;
}
